src/entt/entity/polymorphic.hpp

```diff
@@ -0,0 +1,554 @@
+        return wrapped != other.wrapped;
+    }
+
+    reference operator*() ENTT_NOEXCEPT {
+        Convert convert{};
+        return convert(*wrapped);
+    }
+
+    pointer operator->() ENTT_NOEXCEPT {
+        return std::addressof(operator*());
+    }
+
+private:
+    It wrapped;
+};
+
+
+template<typename PoolsIterator>
+class poly_components_iterator {
+    using iterator_traits = typename std::iterator_traits<PoolsIterator>;
+    using pool_holder_type = typename iterator_traits::value_type;
+
+public:
+    using entity_type = typename pool_holder_type::entity_type;
+    using pointer = std::conditional_t<std::is_const_v<pool_holder_type>, typename pool_holder_type::const_pointer_type, typename pool_holder_type::pointer_type>;
+    using value_type = std::conditional_t<std::is_pointer_v<typename pool_holder_type::value_type>, pointer, constness_as_t<typename pool_holder_type::value_type, pool_holder_type>>;
+    using reference = std::conditional_t<std::is_pointer_v<typename pool_holder_type::value_type>, pointer, value_type&>;
+    using difference_type = typename iterator_traits::difference_type;
+    using iterator_category = std::forward_iterator_tag;
+
+    poly_components_iterator() ENTT_NOEXCEPT = default;
+
+    poly_components_iterator(entity_type e, PoolsIterator pos, PoolsIterator last) ENTT_NOEXCEPT :
+        ent(e), it(pos), end(last) {
+        if (it != end) {
+            current = it->try_get(ent);
+            if(current == nullptr) {
+                ++(*this);
+            }
+        }
+    }
+
+    poly_components_iterator &operator++() ENTT_NOEXCEPT {
+        ++it;
+        while (it != end) {
+            current = it->try_get(ent);
+            if (current != nullptr)
+                break;
+            ++it;
+        }
+        return *this;
+    }
+
+    poly_components_iterator operator++(int) ENTT_NOEXCEPT {
+        poly_components_iterator orig = *this;
+        return ++(*this), orig;
+    }
+
+    bool operator==(const poly_components_iterator& other) const ENTT_NOEXCEPT {
+        return it == other.it;
+    }
+
+    bool operator!=(const poly_components_iterator& other) const ENTT_NOEXCEPT {
+        return it != other.it;
+    }
+
+    pointer operator->() ENTT_NOEXCEPT {
+        return current;
+    }
+
+    reference operator*() ENTT_NOEXCEPT {
+        if constexpr(std::is_pointer_v<value_type>) {
+            return operator->();
+        } else {
+            return *operator->();
+        }
+    }
+
+private:
+    entity_type ent;
+    PoolsIterator it;
+    PoolsIterator end;
+    pointer current = nullptr;
+};
+
+} // namespace entt::internal
+
+/**
+ * Internal details not to be documented.
+ * @endcond
+ */
+
+
+namespace entt {
+
+/** @brief base class for entt::poly_pool_holder */
+template<typename Entity>
+class poly_pool_holder_base {
+public:
+    inline poly_pool_holder_base(basic_sparse_set<Entity>* pool,
```

Comment:
I see, gonna think how to fix this.

Reply:
Can you just turn this:
```cpp
template<typename Entity>
class poly_pool_holder_base
```
Into this:
```cpp
template<typename Storage>
class poly_pool_holder_base
```
Then use like `typename Storage::entity_type` for your purposes? Dunno if it works but the runtime view does something similar to abstract the sparse set.

Reply:
Thank you for your advice. I finally got my hands on looking through runtime view implementation. As I see, it allows to add only storages with the same allocator type. This seems to be similar to my proposal in the comment above, so I will try to implement it and see, how it works. I've also refined the idea a bit, so here it is.

In short, we restrict using pools with different pointer to entity types in the same hierarchy (the pointer type is the only thing we get from allocator, that is used in iteration/contains checks). Then we use allocator type, we got from `storage_traits`. We can do this at compile time for all parent types, or at runtime, only for types, which bind child storages to the `poly_type`. 

In the compile time case, we check, that each parent of a poly type uses allocator with the same pointer type, that's all. In the runtime case the `poly_type<entity_type, value_type>` knows the allocator type of the storage for `value_type` from `storage_traits`, so then we just assert type ids are equal for bound child pools. 

I will probably do a compile time variant first, as it seems more reliable and adds no cost. But runtime one has one advantage - it does not check allocator for components, that were never added.

Reply:
I've implemented the compile time variant, that was mentioned above, however, there are some changes. So right now it looks something like this:

- I've added `entt::poly_type_allocator<Type>` to specialize allocator type for polymorpic component type, by default it uses `std::allocator`. This is only used as default allocator type getter, and in each place, where it is used, user can change it.
- The default specialization of `entt::type_traits` for polymorphic components now uses `entt::poly_type_allocator` to pass allocator type to storage.
- `entt::poly_type<Entity, Type>` now has new template parameter - `Allocator`, and so does `entt::poly_type_holder`, so now they operate on pools using the correct allocator type. When child pool is bound to `entt::poly_type`, it ensures (via `static_assert`), that pointer-to-entity type from the bound pool and from its own allocator type are same.
- When bound as a child pool, the default implementation of `entt::poly_storage_mixin` uses `entt::poly_type_allocator` to provide allocator parameter to `entt::poly_type`, however this can be changed in user defined mixins.
- `entt::poly_types_accessor` specializations now must declare `allocator_type<T>` type meta function, to get the allocator type for the given polymorphic component type. This is required for `entt::assure_poly_type`, which is used in every polymorphic algorithm (and, most likely, in user-defined ones), to pass the allocator type to the accessed `entt::poly_type`. The default specialization of `entt::poly_type_accessor` for `entt::basic_registry`, yet again, uses `entt::poly_type_allocator` for this, but for custom specializations, it gives user complete control over allocator types.